src/models/local_search.rs

```diff
@@ -0,0 +1,405 @@
+//! Local autocompletion search
+//!
+
+use enclose::enclose;
+use futures::FutureExt;
+use http::Request;
+use num::{rational::Ratio, ToPrimitive};
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, DefaultOnError, NoneAsEmptyString};
+use url::Url;
+
+use localsearch::{self, LocalSearch as Searcher, DEFAULT_SCORE_THRESHOLD};
+
+use crate::{
+    constants::{CINEMETA_CATALOGS_URL, CINEMETA_FEED_CATALOG_ID},
+    models::{
+        common::{eq_update, Loadable},
+        ctx::Ctx,
+    },
+    runtime::{
+        msg::{Action, ActionLoad, ActionSearch, Internal, Msg},
+        Effect, EffectFuture, Effects, Env, EnvError, EnvFutureExt, UpdateWithCtx,
+    },
+};
+
+pub use imdb_rating::*;
+
+const INDEX_OPTIONS: IndexOptions = IndexOptions {
+    imdb_rating_weight: 0.5,
+    popularity_weight: 0.5,
+};
+
+/// The response returned when fetching the searchable items list.
+///
+/// Currently this is fetched from Cinemeta's `feed.json`
+#[derive(Deserialize, Serialize, Clone)]
+#[serde(transparent)]
+pub struct SearchableItemsResponse(pub Vec<Searchable>);
+
+#[derive(Copy, Clone)]
+pub struct IndexOptions {
+    imdb_rating_weight: f64,
+    popularity_weight: f64,
+}
+
+/// A searchable item
+#[serde_as]
+#[derive(Clone, PartialEq, Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct Searchable {
+    pub id: String,
+    /// Some deleted or duplicated ids might not have the name
+    /// this is why we default it to empty string and will later
+    /// filter out any items with empty `name` before indexing.
+    #[serde(default)]
+    pub name: String,
+    pub r#type: String,
+    #[serde(default)]
+    #[serde_as(deserialize_as = "DefaultOnError<NoneAsEmptyString>")]
+    pub poster: Option<Url>,
+    #[serde(default)]
+    #[serde_as(deserialize_as = "DefaultOnError<NoneAsEmptyString>")]
+    pub imdb_rating: Option<ImdbRating>,
+    pub popularity: Option<u64>,
+    pub release_info: Option<String>,
+}
+
+/// Local search functionality for the search engine's suggestions when typing
+#[derive(Default, Serialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct LocalSearch {
+    /// The Searchable items that will be used for the local search.
+    ///
+    // #[serde(skip)]
```

Comment:
Commented code?

Reply:
Good catch, this should be uncommented, it was only for testing.